// @codenotary/immudb-node/dist/immu-api/set-val-entry.js
import * as immuGrpc from '../immu-grpc/index.js';
import * as igt from '../immu-grpc-tx/index.js';
import * as igp from '../immu-grpc-precond/index.js';
import * as ike from '../immu-kvm-entry/index.js';
export function createSetValEntries(client) {
    const setGrpc = immuGrpc.unaryCall.createSet(client);
    /**
     * Sets key-value pair(s) for given session defined in credentials
     */
    return function setVEntries(props) {
        return setGrpc({
            request: {
                KVs: props.kvms.map(ike.kvmToGrpcKeyValue),
                noWait: props.options?.dontWaitForIndexer,
                preconditions: props.preconditions?.map(igp.precondToGrpcPrecond),
            },
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('tx must be defined'))
            .then(txGrpc => {
            const txCore = igt.grpcTxHeaderToTxCore(txGrpc);
            const valEntries = props.kvms.map(grpcEntry => ike.kvmToValTxEntry(grpcEntry, txCore.id));
            return { valEntries, txCore };
        });
    };
}
export function createSetValEntriesStreaming(client) {
    const streamSetGrpc = immuGrpc.writerCall.createStreamSet(client);
    /**
     *
     * Example usage:
     *
     * ```ts
     *
     * const tx = setValEntriesStreaming({chunks: fromKVEntries([
     *     {
     *          key: Buffer.from('some key'),
     *          val: Buffer.from('some val'),
     *      },
     *     {
     *          key: Buffer.from('other key'),
     *          val: Buffer.from('other val'),
     *      },
     * ])})
     *
     * ```
     */
    return function setValEntriesStreaming(props) {
        return streamSetGrpc({
            request: props.chunks,
            credentials: props.credentials,
        })
            .then(({ res, info }) => res
            ? res
            : Promise.reject('TxHeader__Output must be defined'))
            .then(igt.grpcTxHeaderToTxCore);
    };
}
