// @codenotary/immudb-node/dist/immu-api/sql-query.js
import * as immuGrpc from '../immu-grpc/index.js';
import * as igs from '../immu-grpc-sql/index.js';
export function createSqlQuery(client) {
    const sqlQueryGrpc = immuGrpc.unaryCall.createSqlQuery(client);
    return function sqlQuery(props) {
        return sqlQueryGrpc({
            request: {
                sql: props.sql,
                params: props.params?.map(igs.sqlNamedValueToGrpcSqlNamedParam),
                reuseSnapshot: props.reuseSnapshot
            },
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('SQLQueryResult__Output must be defined'))
            .then(grpcSqlRows => {
            return igs.grpcQueryResultToListoOfSqlNamedValues(grpcSqlRows);
        });
    };
}
export function createSqlQueryTables(client) {
    const sqlQueryTablesGrpc = immuGrpc.unaryCall.createListTables(client);
    return function sqlQueryTables(props) {
        return sqlQueryTablesGrpc({
            request: {},
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('SQLQueryResult__Output must be defined'))
            .then(grpcSqlRows => {
            return igs.grpcQueryResultToListoOfSqlNamedValues(grpcSqlRows);
        });
    };
}
export function createSqlQueryTable(client) {
    const sqlQueryTableGrpc = immuGrpc.unaryCall.createDescribeTable(client);
    return function sqlQueryTable(props) {
        return sqlQueryTableGrpc({
            request: {
                tableName: props.table,
            },
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('SQLQueryResult__Output must be defined'))
            .then(grpcSqlRows => {
            return igs.grpcQueryResultToListoOfSqlNamedValues(grpcSqlRows);
        });
    };
}
