// @codenotary/immudb-node/dist/immu-api/verification-get-sql-row.js
import * as immuGrpc from '../immu-grpc/index.js';
import * as ver from '../immu-grpc-verification/index.js';
import * as igs from '../immu-grpc-sql/index.js';
export function createGetSqlRowEntryAndVerification(client) {
    const verifiableSqlGetGrpc = immuGrpc.unaryCall.createVerifiableSqlGet(client);
    /**
     *
     */
    return function getSqlRowEntryAndVerification(props) {
        return verifiableSqlGetGrpc({
            request: {
                sqlGetRequest: {
                    atTx: props.dbTxesWindow?.endId,
                    sinceTx: props.dbTxesWindow?.startId,
                    pkValues: props.pk.map(igs.sqlValueToGrpcSqlValue)
                },
                proveSinceTx: props.refTxId,
            },
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('VerifiableTx__Output must be defined'))
            .then(grpcVerSql => {
            const grpcVerTx = grpcVerSql.verifiableTx;
            if (grpcVerTx == undefined) {
                throw 'verifiableTx must be defined';
            }
            const grpcTx = grpcVerTx.tx;
            const grpcProof = grpcVerTx.dualProof;
            if (grpcTx?.header == undefined) {
                throw 'transaction must be defined';
            }
            if (grpcProof?.sourceTxHeader == undefined) {
                throw 'sourceTxHeader must be defined';
            }
            if (grpcProof?.targetTxHeader == undefined) {
                throw 'targetTxHeader must be defined';
            }
            const grpcTxHeader = grpcTx.header;
            const verificationEntries = ver.grpcTxEntriesToVerificationEntries({
                txEntries: grpcTx.entries,
                allEntriesMht: grpcTxHeader.eH,
            });
            const verificationTx = ver.grpcDualProofToVerificationTx({
                grpcProof: grpcProof,
                grpcTx: grpcTxHeader,
                refHash: props.refHash,
                refTxId: props.refTxId,
            });
            const verification = {
                entries: verificationEntries,
                tx: verificationTx
            };
            const transaction = {
                id: verificationTx.tx.id,
                timestamp: verificationTx.tx.timestamp,
                entries: verificationEntries.allEntries,
            };
            return {
                transaction,
                verification
            };
        });
    };
}
