// @codenotary/immudb-node/dist/immu-api/verification-set-ref.js
import * as immuGrpc from '../immu-grpc/index.js';
import * as igp from '../immu-grpc-precond/index.js';
import * as ver from '../immu-grpc-verification/index.js';
import Long from 'long';
export function createSetRefEntryGetVerification(client) {
    const grpcSetRefAndGetVerification = immuGrpc.unaryCall.createVerifiableSetReference(client);
    /**
     * Sets key-value pair(s) for given session defined in credentials
     */
    return function setRefEntryGetVerification(props) {
        return grpcSetRefAndGetVerification({
            request: {
                referenceRequest: {
                    referencedKey: props.referToKey,
                    key: props.key,
                    atTx: props.keyTxId,
                    boundRef: props.boundRef,
                    preconditions: props.preconditions?.map(igp.precondToGrpcPrecond),
                    noWait: props.options?.dontWaitForIndexer,
                },
                proveSinceTx: props.refTxId,
            },
            options: {
                credentials: props.credentials,
            },
        })
            .then(maybeResponse => maybeResponse
            ? maybeResponse
            : Promise.reject('VerifiableTx__Output must be defined'))
            .then(grpcVerTx => {
            const grpcTx = grpcVerTx.tx;
            const grpcProof = grpcVerTx.dualProof;
            if (grpcTx?.header == undefined) {
                throw 'transaction must be defined';
            }
            if (grpcProof?.sourceTxHeader == undefined) {
                throw 'sourceTxHeader must be defined';
            }
            if (grpcProof?.targetTxHeader == undefined) {
                throw 'targetTxHeader must be defined';
            }
            const grpcTxHeader = grpcTx.header;
            const refEntry = {
                type: 'ref',
                version: '1',
                key: props.key,
                referredKey: props.referToKey,
                referredAtTxId: props.keyTxId ?? Long.UZERO
            };
            const verificationEntries = {
                type: 'all-of',
                allEntries: [refEntry],
                allEntriesMht: grpcTxHeader.eH,
            };
            const verificationTx = ver.grpcDualProofToVerificationTx({
                grpcProof: grpcProof,
                grpcTx: grpcTxHeader,
                refHash: props.refHash,
                refTxId: props.refTxId,
            });
            const verification = {
                entries: verificationEntries,
                tx: verificationTx
            };
            const transaction = {
                id: verificationTx.tx.id,
                timestamp: verificationTx.tx.timestamp,
                entries: verificationEntries.allEntries,
            };
            return {
                transaction,
                verification,
                refEntry,
            };
        });
    };
}
