// @codenotary/immudb-node/dist/common/errors.js
export class ImmudbClientError extends Error {
    constructor(message) {
        super(message);
        // Ensure the name of this error is the same as the class name
        this.name = this.constructor.name;
        // This clips the constructor invocation from the stack trace. It's not
        // absolutely essential, but it does make the stack trace a little
        // nicer.
        // https://nodejs.org/api/errors.html#errors_error_capturestacktrace_targetobject_constructoropt
        Error.captureStackTrace(this, this.constructor);
    }
}
export class InternalError extends ImmudbClientError {
    constructor(err, info) {
        super(err.message);
        this.data = { err, info };
    }
}
export class NoResponseValueError extends ImmudbClientError {
    constructor(message, info) {
        super(message);
        this.data = info;
    }
}
export class OperationCanceledError extends ImmudbClientError {
    constructor(message, info) {
        super(message);
        this.data = info;
    }
}
