// @codenotary/immudb-node/dist/immu-permission.js
export function permissionToKnownIfPossible(permission) {
    if (typeof permission === 'string') {
        return permission;
    }
    return permissionFromCode(permission);
}
export function permissionFromCode(code) {
    const map = knownPermissionCodeToName;
    return map[code] ?? code;
}
export function permissionToCode(permission) {
    return typeof permission === 'number'
        ? permission
        : knownPermissionNameToCode[permission];
}
/**
 * Permission names to codes.
 */
export const knownPermissionNameToCode = {
    /**
     * No permissions
     */
    None: 0,
    /**
     *Permission to read database
     */
    Read: 1,
    /**
     * Permission to read and write database
     */
    ReadWrite: 2,
    /**
     * Permission to read, write and administer database
     */
    Admin: 254,
    /**
     * Permission to read, write and administer all databases
     */
    SysAdmin: 255,
};
export const knownPermissionCodeToName = {
    /**
     * No permissions
     */
    0: 'None',
    /**
     *Permission to read database
     */
    1: 'Read',
    /**
     * Permission to read and write database
     */
    2: 'ReadWrite',
    /**
     * Permission to read, write and administer database
     */
    254: 'Admin',
    /**
     * Permission to read, write and administer all databases
     */
    255: 'SysAdmin',
};
