// @codenotary/immudb-node/dist/immu-grpc/reader-call.js
import * as promis from '../grpc-promis/index.js';
// **************************
// Get Values
// **************************
export function createStreamGet(client) {
    return promis.promisifyGrpcReaderAsync(client.streamGet.bind(client));
}
export function createStreamScan(client) {
    return promis.promisifyGrpcReaderAsync(client.streamScan.bind(client));
}
export function createStreamZScan(client) {
    return promis.promisifyGrpcReaderAsync(client.streamZScan.bind(client));
}
export function createStreamHistory(client) {
    return promis.promisifyGrpcReaderAsync(client.streamHistory.bind(client));
}
// **************************
// Get Values and proof
// **************************
export function createStreamVerifiableGet(client) {
    return promis.promisifyGrpcReaderAsync(client.streamVerifiableGet.bind(client));
}
// **************************
// Instance
// **************************
export function createExportTx(client) {
    return promis.promisifyGrpcReaderAsync(client.exportTx.bind(client));
}
