// @codenotary/immudb-node/dist/immu-grpc/unary-call.js
import * as promis from '../grpc-promis/index.js';
// **************************
// Session
// **************************
export function createOpenSession(client) {
    return promis.promisifyGrpcCall(client.openSession.bind(client));
}
export function createCloseSession(client) {
    return promis.promisifyGrpcCall(client.closeSession.bind(client));
}
export function createKeepAlive(client) {
    return promis.promisifyGrpcCall(client.keepAlive.bind(client));
}
// **************************
// SQL transactional
// **************************
export function createNewTx(client) {
    return promis.promisifyGrpcCall(client.newTx.bind(client));
}
export function createCommit(client) {
    return promis.promisifyGrpcCall(client.commit.bind(client));
}
export function createRollback(client) {
    return promis.promisifyGrpcCall(client.rollback.bind(client));
}
export function createTxSqlExec(client) {
    return promis.promisifyGrpcCall(client.txSqlExec.bind(client));
}
export function createTxSqlQuery(client) {
    return promis.promisifyGrpcCall(client.txSqlQuery.bind(client));
}
// **************************
// SQL
// **************************
export function createSqlExec(client) {
    return promis.promisifyGrpcCall(client.sqlExec.bind(client));
}
export function createSqlQuery(client) {
    return promis.promisifyGrpcCall(client.sqlQuery.bind(client));
}
// **************************
// SQL util
// **************************
export function createListTables(client) {
    return promis.promisifyGrpcCall(client.listTables.bind(client));
}
export function createDescribeTable(client) {
    return promis.promisifyGrpcCall(client.describeTable.bind(client));
}
// **************************
// SQL and proof
// **************************
export function createVerifiableSqlGet(client) {
    return promis.promisifyGrpcCall(client.verifiableSqlGet.bind(client));
}
// **************************
// Get Values
// **************************
export function createGet(client) {
    return promis.promisifyGrpcCall(client.get.bind(client));
}
export function createGetAll(client) {
    return promis.promisifyGrpcCall(client.getAll.bind(client));
}
export function createTxById(client) {
    return promis.promisifyGrpcCall(client.txById.bind(client));
}
export function createScan(client) {
    return promis.promisifyGrpcCall(client.scan.bind(client));
}
export function createTxScan(client) {
    return promis.promisifyGrpcCall(client.txScan.bind(client));
}
export function createZScan(client) {
    return promis.promisifyGrpcCall(client.zScan.bind(client));
}
export function createHistory(client) {
    return promis.promisifyGrpcCall(client.history.bind(client));
}
// **************************
// Get Values and proof
// **************************
export function createVerifiableGet(client) {
    return promis.promisifyGrpcCall(client.verifiableGet.bind(client));
}
export function createVerifiableTxById(client) {
    return promis.promisifyGrpcCall(client.verifiableTxById.bind(client));
}
// **************************
// Set Values
// **************************
export function createSet(client) {
    return promis.promisifyGrpcCall(client.set.bind(client));
}
export function createSetReference(client) {
    return promis.promisifyGrpcCall(client.setReference.bind(client));
}
export function createZAdd(client) {
    return promis.promisifyGrpcCall(client.zAdd.bind(client));
}
export function createExecAll(client) {
    return promis.promisifyGrpcCall(client.execAll.bind(client));
}
export function createDelete(client) {
    return promis.promisifyGrpcCall(client.delete.bind(client));
}
// **************************
// Set Values and proof
// **************************
export function createVerifiableSet(client) {
    return promis.promisifyGrpcCall(client.verifiableSet.bind(client));
}
export function createVerifiableSetReference(client) {
    return promis.promisifyGrpcCall(client.verifiableSetReference.bind(client));
}
export function createVerifiableZAdd(client) {
    return promis.promisifyGrpcCall(client.verifiableZAdd.bind(client));
}
// **************************
// User
// **************************
export function createCreateUsers(client) {
    return promis.promisifyGrpcCall(client.createUser.bind(client));
}
export function createListUsers(client) {
    return promis.promisifyGrpcCall(client.listUsers.bind(client));
}
export function createChangePassword(client) {
    return promis.promisifyGrpcCall(client.changePassword.bind(client));
}
export function createChangePermission(client) {
    return promis.promisifyGrpcCall(client.changePermission.bind(client));
}
export function createSetActiveUser(client) {
    return promis.promisifyGrpcCall(client.setActiveUser.bind(client));
}
// **************************
// Database
// **************************
export function createDatabaseListV2(client) {
    return promis.promisifyGrpcCall(client.databaseListV2.bind(client));
}
export function createGetDatabaseSettingsV2(client) {
    return promis.promisifyGrpcCall(client.getDatabaseSettingsV2.bind(client));
}
export function createCreateDatabaseV2(client) {
    return promis.promisifyGrpcCall(client.createDatabaseV2.bind(client));
}
export function createUpdateDatabaseV2(client) {
    return promis.promisifyGrpcCall(client.updateDatabaseV2.bind(client));
}
export function createUseDatabase(client) {
    return promis.promisifyGrpcCall(client.useDatabase.bind(client));
}
export function createLoadDatabase(client) {
    return promis.promisifyGrpcCall(client.loadDatabase.bind(client));
}
export function createUnloadDatabase(client) {
    return promis.promisifyGrpcCall(client.unloadDatabase.bind(client));
}
export function createCompactIndex(client) {
    return promis.promisifyGrpcCall(client.compactIndex.bind(client));
}
export function createFlushIndex(client) {
    return promis.promisifyGrpcCall(client.flushIndex.bind(client));
}
export function createDeleteDatabase(client) {
    return promis.promisifyGrpcCall(client.deleteDatabase.bind(client));
}
export function createDatabaseHealth(client) {
    return promis.promisifyGrpcCall(client.databaseHealth.bind(client));
}
// **************************
// Instance
// **************************
export function createHealth(client) {
    return promis.promisifyGrpcCall(client.health.bind(client));
}
export function createCurrentState(client) {
    return promis.promisifyGrpcCall(client.currentState.bind(client));
}
